### PR TITLE
chore: align `Headers` to spec

### PR DESCRIPTION
**Pull request by lucacasonato**
This commit aligns `Headers` to spec. It also removes the now unused
03_dom_iterable.js file. We now pass all relevant `Headers` WPT. We do
not implement any sort of header filtering, as we are a server side
runtime.

This is likely not the most efficient implementation of `Headers` yet. It is
however spec compliant. Once all the APIs in the `HTTP` hot loop are
correct we can start optimizing them. It is likely that this reduces bench
throughput temporarily.

### CODE REVIEW

File: op_crates/web/00_infra.js [hunk 1]
```diff
@@ -8,6 +8,74 @@
...
+        return `\\u${char.charCodeAt(0).toString(16).padStart(4, "0")}-\\u${
+          char.charCodeAt(2).toString(16).padStart(4, "0")
+        }`;
+      } else {
+        throw TypeError("unreachable");
+      }
+    });
+    return matchers.join("");
+  }
+
```
Comment (lucacasonato) [inline op_crates/web/00_infra.js:78]: This is all just moved from //op_crates/web/01_mimesniff.js

File: tools/wpt/expectation.json [hunk 2]
```diff
@@ -684,6 +684,15 @@
...
         ]
+      },
+      "headers": {
+        "headers-basic.any.js": true,
+        "headers-casing.any.js": true,
+        "headers-combine.any.js": true,
+        "headers-errors.any.js": true,
+        "headers-normalize.any.js": true,
+        "headers-record.any.js": true,
+        "headers-structure.any.js": true
```
Comment (lucacasonato) [inline tools/wpt/expectation.json:695]: All relevant tests pass.

**Comment by lucacasonato** [comment]
The `unit_test/fetch.ts` changes are harmless and not technically wrong, but not great. I can not easily resolve this without conflicting with my follow-up PR that rewrites `Body`, and `Request`, so I would vote to keep it as is for now and revert the test changes in the follow up PR.

File: cli/tests/unit/headers_test.ts [hunk 3]
```diff
@@ -372,22 +372,12 @@ unitTest(function headersAppendDuplicateSetCookieKey(): void {
...
-unitTest(function headersSetDuplicateCookieKey(): void {
-  const headers = new Headers([["Set-Cookie", "foo=bar"]]);
-  headers.set("set-Cookie", "foo=baz");
-  headers.set("set-cookie", "bar=qat");
-  const actual = [...headers];
-  assertEquals(actual, [
-    ["set-cookie", "foo=baz"],
-    ["set-cookie", "bar=qat"],
-  ]);
-});
```
Comment (lucacasonato) [inline cli/tests/unit/headers_test.ts:389]: This test is plain wrong. `headers.set` is not the same as append. It removes existing headers of that name before insertion. See https://fetch.spec.whatwg.org/#concept-header-list-set

**Review by ry — APPROVED** [review]
LGTM

**Comment by ry** [comment]
This PR cause a fairly severe performance degradation in the performance of std/http/server.ts

Screen Shot 2021-05-06 at 7 57 20 AM

**Comment by lucacasonato** [comment]
This is most likely cause by the O(n^2) complexity of https://github.com/denoland/deno/blob/main/extensions/fetch/20_headers.js#L178-L207

**Pull request merged**